### DISCUSSION TITLE
Can i create or use (from external crate validator) custom validation for fields in input objects

### DISCUSSION THREAD

**Discussion by WrldEngine** [graphql-rust/juniper#1296] [Q&A] [answered]
For example:
```rust
#[derive(juniper::GraphQLInputObject, Clone)]
pub struct UserInput {
    pub id: i32,
    #[validate(email)]
    pub email: String,
    #[validate(length(min = 5, max = 150), custom(function = "validate_username"))]
    pub username: String,
    pub password: String,
}
```
Because, for me, is not comfortable to make my own implementation/macro manually
I couldnt find any information about this, so i hope i can find solution here

**Comment by tyranron** [comment, marked as answer]
@WrldEngine well, you may use the [`validator` crate](https://docs.rs/validator) to do exactly what you've described, and then, whenever you receive such an input object in your GraphQL resolvers, just call the `.validate()` and return the errors, if any.

However, the idiomatic way to deal with this kind of stuff is to ["parse, don't validate"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/). For that, we have [scalars in GraphQL](https://graphql-rust.github.io/juniper/types/scalars.html). Instead of just using `String` for your `email`, `username` and `password`, it's better to introduce newtypes `Email`, `Username` and `Password`, implementing `GraphQLScalar` and, of course, containing already valid values only after parsing from an input.

For `password` particularly, you especially need a custom `GraphQLScalar` which internally utilizes the [`secrecy` crate](https://docs.rs/secrecy), since operating with credentials plainly is not a good practice, leading to leaking them into logs, debug output, etc.

**Reply by WrldEngine** [reply]
@tyranron Unfortunately, i already solve my problem, the exactly as you wrote
Because i didn't read docs attentively
So, i decided to close this discussion
Anyway, thank you for your responsibility
